File: i2c.py
import smbus
import time

class I2CUtils:
    def __init__(self):
        self.i2c_allow = False
        self.is_training_active = False
        self.n_devices = 0
        self.i2c_addr = []
        self.img_id = 1
        self.resp = 0
        self.CMD_SHOW = 6
        self.CMD_BLACK = 3
        self.CMD_RESET = 4
        self.CMD_IMG = 5
        self.RIGHT_M0_ADDR = 0x04  
        self.LEFT_M0_ADDR = 0x14   
        self.dir_str = ['L', 'R']
        self.correct_dir = None
        self.bus = smbus.SMBus(1)  # 1 indicates /dev/i2c-1

    def setup_i2c(self):
        self.i2c_allow = True
        self.i2c_scanner()

    def i2c_scanner(self):
        print("Scanning for I2C devices...")
        self.n_devices = 0
        for address in range(0x03, 0x78):
            try:
                self.bus.write_quick(address)
                print(f"I2C device found at address 0x{address:02X}!")
                self.i2c_addr.append(address)
                self.n_devices += 1
            except IOError:
                pass  # Address is not in use
        if self.n_devices == 0:
            print("No I2C devices found")
        else:
            print("Scanning complete")

    def receive_event(self):
        try:
            data = self.bus.read_byte(self.RIGHT_M0_ADDR)  
            print(f"Received data: {data}")
        except Exception as e:
            print(f"Error receiving event: {e}")

    def send_cmd(self, cmd, m0_id):
        try:
            self.bus.write_byte(m0_id, cmd)
            print(f"Command {cmd} sent to device {m0_id}")
        except Exception as e:
            print(f"Failed to send command: {e}")

    def response_from_right_m0(self):
        self.resp = 0x00
        while self.resp == 0x00:
            try:
                self.resp = self.bus.read_byte(self.RIGHT_M0_ADDR)
                time.sleep(0.005)
                print("Acknowledgment from Right M0")
            except IOError:
                pass  

    def response_from_left_m0(self):
        self.resp = 0x00
        while self.resp == 0x00:
            try:
                self.resp = self.bus.read_byte(self.LEFT_M0_ADDR)
                time.sleep(0.01)
                print("Acknowledgment from Left M0")
            except IOError:
                pass 

    def send_black_to_all_m0s(self):
        for address in self.i2c_addr:
            self.send_black(address)
            # Wait for acknowledgment
            self.resp = 0x00
            while self.resp == 0x00:
                try:
                    self.resp = self.bus.read_byte(address)
                    time.sleep(0.01)
                    print(f"Black Response from M0 at address {address}")
                except IOError:
                    pass  # Keep waiting for acknowledgment

    def send_image_to_m0(self, img_id_str, m0_id):
        img_data = [ord(char) for char in img_id_str]
        try:
            self.bus.write_i2c_block_data(m0_id, self.CMD_IMG, img_data)
            print(f"Image ID '{img_id_str}' sent to M0 {m0_id}")
        except Exception as e:
            print(f"Failed to send image: {e}")

    def send_black(self, m0_id):
        self.send_cmd(self.CMD_BLACK, m0_id)

    def stop(self):
        self.bus.close()

File: controller/I2C.cpp
#include "I2C.h"
#define DEBUG

//#define MAX_IMAGES 20 

byte i2cAddr[8];
int nDevices = 0;
unsigned long lastI2CScanMs, i2cScanDurationMs = 10000;
bool i2c_allow = false;
bool isTrainingActive= false;
byte img_id = 1;
int resp;
// TODO: Make these enum
int CMD_SHOW = 6;
int CMD_BLACK = 3;
int CMD_RESET = 4;
int CMD_IMG = 5;
//Choice_dir choiceDir;
const char dirStr[3] = {'L', 'R'};
char correctDir;


void setupI2C() {
    i2c_allow = true;
    Wire.begin();
    i2cScanner();
}



void i2cScanner() {
  // https://learn.adafruit.com/scanning-i2c-addresses/arduino
   byte error, address;

  Serial.println("Scanning...");

  nDevices = 0;
  for(address = 1; address < 8; address++ ) 
  {

    Wire.beginTransmission(address);
    error = Wire.endTransmission();

    if (error == 0)
    {
      Serial.print("I2C device found at address 0x");
      if (address<16) 
        Serial.print("0");
      Serial.print(address,HEX);
      Serial.println("  !");

      i2cAddr[nDevices] = address;

      nDevices++;
    }
    else if (error==4) 
    {
      Serial.print("Unknown error at address 0x");
      if (address<16) 
        Serial.print("0");
      Serial.println(address,HEX);
    }    
  }
  if (nDevices == 0)
    Serial.println("No I2C devices found\n");
  else
    Serial.println("done\n");

  delay(500);
}
//
void receiveEvent(int howMany)
{
  if(0 < Wire.available()) {
    int x = Wire.read();    // receive byte as an integer
    Serial.println(x);

  }
}

void sendCmd(int cmd, int m0_id){
  Wire.beginTransmission(m0_id);
  Wire.write(cmd);
  Wire.endTransmission();
  //Serial.println(String(cmd) + " sent to M0 " + String(m0_id));
}



void ResponseFromRightM0(){
    resp = 0x00;
  while(resp == 0x00) {

    Wire.requestFrom(RIGHT_M0_ADDR , 1);
    if(Wire.available() > 0) {
      resp = Wire.read();
      delay(5);
      Serial.println("Acknowledgement from M0 2");
break;
    }
  }

}


void ResponseFromLeftM0(){
    resp = 0x00;
  while(resp == 0x00) {

    Wire.requestFrom(LEFT_M0_ADDR, 1);
    if(Wire.available() > 0) {
      resp = Wire.read();
      delay(10);
      Serial.println("Acknowledgement from M0 4");
break;
    }
  }

}


void sendBlackToAllM0s(void) {
    for(int i=0; i<nDevices; i++)
      {
        send_black(i2cAddr[i]);
        }
//        for(int i = 0; i < nDevices; i++) { 
//        // keep waiting until get a response rather than 0x00 from m0
//        resp = 0x00;
//        while(resp == 0x00)
//        {
//        Wire.beginTransmission(i2cAddr[i]);
//        Wire.requestFrom(i2cAddr[i], 1);
//          if(Wire.available() > 0)
//          {
//              resp = Wire.read();
//            
//              delay(10);
//              Serial.println("Black Response from M0 " + String(i2cAddr[i]));
//
//          }
//        }
//        Wire.endTransmission();
//      }

}


void i2c_send_img(String img_id, int m0_id, int len)
{
  byte data[len];
  img_id.getBytes(data, len);
  for(int i=0; i<len; i++)
  //Serial.println("Byte " + String(i) + ": " + data[i]);  
  Wire.beginTransmission(m0_id);
  Wire.write(data, len);
  Wire.endTransmission();
  Serial.println("IMG: " + img_id + " sent to M0 " + String(m0_id));
}



void send_black(byte m0_id)
{
  Wire.beginTransmission(m0_id);
  Wire.write(CMD_BLACK);
  Wire.endTransmission();
  //Serial.println("Black command sent to M0 " + String(m0_id));
}
